Add support for cli metadata export

sandhya-nayak opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our [contributor guidelines](https://github.com/elyra-ai/community/blob/master/contributing.md)
    1.1 Please follow the [7 rules for a great commit message](https://github.com/elyra-ai/community/blob/master/contributing.md#creating-a-pull-request)
  2. If the PR is unfinished, leave it as 'DRAFT', add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...' or use the 'status:Work in Progress' label.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If this PR involves UI changes, please provide a screen capture (before/after) for a faster review.
  6. Remember to add 'Fixes #ISSUE_NUMBER' (or any [valid keyword](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)) to the end of your message body to get the issue properly closed when the PR is merged.
  7. Set the proper milestone based on the target release for the issue.
  8. Consider whether any documentation need to be added or updated based on your changes.
-->

### What changes were proposed in this pull request?
Fixes #2404.

1. Add support for metadata export in the cli
2. Add tests for export subcommand
3. Update documentation
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor code that leads to class changes, showing the updated class hierarchy will help reviewers.
  2. If there is design documentation, please add the link.
-->

### How was this pull request tested?
Ran make clean, make install, make test and make docs to verify that the installation and tests were successful and documentation was being rendered correctly.
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly
including negative and positive cases if possible.

If it was tested in a way different from regular unit tests, please clarify how you tested step by step,
ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.

If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

kevin-bates reviewed (commented):
This looks and works great - thank you @sandhya-nayak!  I just had a couple of minor comments/suggestions.

sandhya-nayak commented on elyra/metadata/metadata_app.py:
```diff
@@ -376,6 +378,76 @@ def start(self):
+            os.makedirs(dest_directory)
+
+        if clean:
+            print("Cleaning out all files in {}".format(dest_directory))
+            files = [os.path.join(dest_directory, f) for f in os.listdir(dest_directory)]
+            [os.remove(f) for f in files if os.path.isfile(f)]
```
will do 👍

sandhya-nayak replied:
Added code to make the print statement conditional on the presence of files. Please review @kevin-bates.

sandhya-nayak commented on elyra/metadata/metadata_app.py:
```diff
@@ -376,6 +378,76 @@ def start(self):
+            files = [os.path.join(dest_directory, f) for f in os.listdir(dest_directory)]
+            [os.remove(f) for f in files if os.path.isfile(f)]
+
+        print("Exporting metadata instances for schemaspace {}".format(self.schemaspace) +
+              (" and schema {}".format(schema_name) if schema_name else "") +
+              " to {}".format(dest_directory))
+        for metadata in metadata_instances:
+            dict_metadata = metadata.to_dict()
+            output_file = os.path.join(dest_directory, f'{dict_metadata["name"]}.json')
+            with open(output_file, mode='w') as output_file:
+                json.dump(dict_metadata, output_file, indent=4)
```
Individual logging - so we know how many and which ones were exported - alright I can make that change. 👍

sandhya-nayak replied:
Added code for individual logging of exported metadata (as in case of `list` subcommand). Added an additional column for the export location.

kevin-bates replied:
@ptitzler - we're contradicting each other (sorry @sandhya-nayak!):
> Not sure if we want to list each instance (this can give users a "warm fuzzy") but I'm okay not for now. One thing we could do in that case is add an indication that the instance is invalid (like is done in list). Hmm, since the inclusion of invalid instances is the default, individual logging might be warranted.

I'm okay only producing a count, although the invalid instances are where things get funky for me.  Perhaps a counter like:

`5 instances exported (2 of which are invalid)`

ptitzler replied:
> ... though the invalid instances are where things get funky for me

How so? What would a user do with this extra information?

kevin-bates replied:
For starters, they would know their export contains invalid instances and which ones are invalid.

Seems like `import` would need an `ignore-invalid` option.

Btw, I've always felt like `--valid-only` was an odd option and even odder default (that invalid instances are included by default).  I think it was because we wanted to have the existence of invalid instances more well-known to the user (thus the default).  But that's a different topic.

ptitzler replied:
Ah, I see, so the default behavior. Well, that could be changed in 4.0 (breaking change). I wasn't sure wether the message meant to suggest to the user that some action needs to be taken (e.g. use the `--valid-only` option to ignore the invalid ones, which I agree should not be imported by default.)

ptitzler commented on docs/source/user_guide/code-snippets.md:
````diff
@@ -122,6 +122,24 @@ elyra-metadata install code-snippets \
+
+```bash
+elyra-metadata export code-snippets \
+	--schema-name="code-snippet" \
````
```suggestion
	--schema_name="code-snippet" \
```

sandhya-nayak replied:
uh oh - good catch - I'll fix it.

ptitzler replied:
Since there's only one schemaspace in the `code-snippets` namespace, we should not display the `--schema_name="code-snippets"`  option in the example and the explanatory text below.

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler commented on docs/source/user_guide/code-snippets.md:
```diff
@@ -122,6 +122,24 @@ elyra-metadata install code-snippets \
+
+The above example will export all code snippets in the "code-snippet" schema to the "/tmp/foo/code-snippets" directory.
+
+Note that you must specify the `--directory` option. If the `--schema-name` option is not specified, code snippets across all schemas will be exported. 
```
```suggestion
Note that you must specify the `--directory` option. If the `--schema_name` option is not specified, code snippets across all schemas will be exported. 
```

sandhya-nayak replied:
Looks like I've made this mistake in multiple places - I'll do a find and replace so I'll get them all.

ptitzler replied:
Since there's only one schemaspace in the `code-snippets` namespace, we should not display the `--schema_name="code-snippets"`  option in the example and the explanatory text below.

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler commented on docs/source/user_guide/pipeline-components.md:
````diff
@@ -192,6 +192,24 @@ Note: You must specify all property values, not only the ones that you want to m
+
+```bash
+elyra-metadata export component-catalogs \
+	--schema-name="url-catalog" \
````
```suggestion
	--schema_name="url-catalog" \
```

ptitzler replied:
While completely correct (there are different schema names defined in this namespace), we should keep it simple and not mention the `--schema_name` option in the example or text below. Given that we are positioning export as the means to create a backup, I don't believe it's likely that most users would want to export, for example, only URL catalog entries. The built-in command help calls out that it is possible to filter by schema name.

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler commented on docs/source/user_guide/pipeline-components.md:
```diff
@@ -192,6 +192,24 @@ Note: You must specify all property values, not only the ones that you want to m
+
+The above example will export all component catalogs in the "url-catalog" schema to the "/tmp/foo/component-catalogs" directory.
+
+Note that you must specify the `--directory` option. If the `--schema-name` option is not specified, component catalogs across all schemas will be exported. 
```
```suggestion
Note that you must specify the `--directory` option. If the `--schema_name` option is not specified, component catalogs across all schemas will be exported. 
```

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler commented on docs/source/user_guide/runtime-conf.md:
````diff
@@ -122,6 +122,24 @@ elyra-metadata install runtimes \
+
+```bash
+elyra-metadata export runtimes \
+	--schema-name="kfp" \
````
```suggestion
	--schema_name="kfp" \
```

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler commented on docs/source/user_guide/runtime-conf.md:
```diff
@@ -122,6 +122,24 @@ elyra-metadata install runtimes \
+
+The above example will export all runtime configurations in the "kfp" schema to the "/tmp/foo/runtimes" directory.
+
+Note that you must specify the `--directory` option. If the `--schema-name` option is not specified, runtime configurations across all schemas will be exported. 
```
```suggestion
Note that you must specify the `--directory` option. If the `--schema_name` option is not specified, runtime configurations across all schemas will be exported. 
```

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler commented on docs/source/user_guide/runtime-image-conf.md:
````diff
@@ -118,6 +118,24 @@ $ elyra-metadata install runtime-images --schema_name=runtime-image \
+
+```bash
+elyra-metadata export runtime-images \
+	--schema-name="runtime-image" \
````
```suggestion
	--schema_name="runtime-image" \
```

ptitzler replied:
Since there's only one schemaspace in the `runtime-images` namespace, we should not display the `--schema_name="runtime-image"`  option in the example and the explanatory text below.

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler commented on docs/source/user_guide/runtime-image-conf.md:
```diff
@@ -118,6 +118,24 @@ $ elyra-metadata install runtime-images --schema_name=runtime-image \
+
+The above example will export all runtime image configurations in the "runtime-image" schema to the "/tmp/foo/runtime-images" directory.
+
+Note that you must specify the `--directory` option. If the `--schema-name` option is not specified, runtime image configurations across all schemas will be exported. 
```
```suggestion
Note that you must specify the `--directory` option. If the `--schema_name` option is not specified, runtime image configurations across all schemas will be exported. 
```

ptitzler replied:
See above (remove `--sche...` reference to keep it simple.

sandhya-nayak replied:
Removed all mentions of `--schema-name` parameter from the documentation as discussed via chat.

ptitzler reviewed (changes requested):
The following command 
```
$ elyra-metadata export component-catalogs --directory=/tmp/foo2 --schema_name=MYBOGUSSCHEMA --clean
```

produces invalid output:
```
...
Creating directory structure for /tmp/foo2/component-catalogs
Cleaning out all files in /tmp/foo2/component-catalogs
Exporting metadata instances for schemaspace component-catalogs and schema MYBOGUSSCHEMA to /tmp/foo2/component-catalogs
```

```
$ ls -al /tmp/foo2/component-catalogs
total 16
drwxr-xr-x  4 patti  wheel  128 Feb  4 14:24 .
drwxr-xr-x  3 patti  wheel   96 Feb  4 14:24 ..
-rw-r--r--  1 patti  wheel  618 Feb  4 14:24 no_such_airflow_archive.json
-rw-r--r--  1 patti  wheel  700 Feb  4 14:24 no_such_archive.json
```

I'm suspecting this is caused by the fact that I have invalid instances defined, which are exported, even though they have a schema name other than `BOGUS...`

ptitzler commented on elyra/metadata/metadata_app.py:
```diff
@@ -376,6 +378,76 @@ def start(self):
+            metadata_instances = None
+
+        if not metadata_instances:
+            print("No metadata instances found for schemaspace {}".format(self.schemaspace) +
```
[Code style] Although we haven't been consistent in the past, we have started to 
- surround identifiers with single quotes to make it easier to read messages
- use f-strings instead of '...'.format(...)

Could you please update your code to follow the (undocumented) convention? 
```
print("No metadata instances found for schemaspace {}".format(self.schemaspace) +
```
 
to 
```
print(f"No metadata instances found for schemaspace '{self.schemaspace}'" ...
```
```suggestion
            print("No metadata instances found for schemaspace {}".format(self.schemaspace) +
```

sandhya-nayak replied:
will do 👍

sandhya-nayak replied:
Updated print statements to use f-strings instead of format and surrounded identifiers with single quotes.

sandhya-nayak commented:
Added new changes addressing the following:

1. When `--clean` parameter is passed, the statement "Cleaning all files in the <export_directory>" is now displayed only if there are files to clean up. (requested by @kevin-bates)
2. The exported files are listed out on the console, in a manner similar to the output of the `list` subcommand, with the addition of an `Export Location` column. (requested by @kevin-bates)
3. Removed all mentions of the `--schema_name` parameter from the documentation as suggested by @ptitzler.
4. The [issue](#pullrequestreview-873644287) that Patrick highlighted required a fix in manager.py. In the `get_all` method, if an exception is caught (instance is invalid), a copy of the bad instance is being appended to the output list of instances. Previously, this instance would be appended only if `include_invalid` was `True`. There was no condition based on the `of_schema` value and therefore, all invalid instances irrespective of schema were being returned. The condition was updated to only append the invalid instance to the output list of instances if `of_schema` is `False` or if `of_schema` matches the instance's schema.

ptitzler commented:
```
$ elyra-metadata export component-catalogs --clean --directory=/tmp/foo3 --valid-only --schema_name=blah
Schema name 'blah' is invalid. For the 'component-catalogs' schemaspace, the schema-name must be one of ['airflow-package-catalog', 'airflow-provider-package-catalog', 'local-directory-catalog', 'local-file-catalog', 'url-catalog']
Traceback (most recent call last):
  File "/opt/anaconda3/envs/cli-metadata-export/bin/elyra-metadata", line 8, in <module>
    sys.exit(MetadataApp.main())
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 603, in main
    elyra_metadata.start()
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 624, in start
    return subinstance.start()
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 524, in start
    return subinstance.start()
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 417, in start
    raise SchemaNotFoundError(self.schemaspace, schema_name)
elyra.metadata.error.SchemaNotFoundError: No such schema named 'blah' was found in the component-catalogs schemaspace.
```

- We should not dump the stacktrace because it does not contain any useful information
- if possible, only the first error message (`Schema name 'blah' is invalid. For the 'component-catalogs'  ...`) because the second one is redundant
- possible typo in first message `schema-name`: should be `schema name` or `schema_name`

ptitzler commented:
We need to catch I/O errors to prevent stacktrace dumps and surface the raising exception in a more user friendly way (e.g. `Error creating directory structure: <caught-exception-text>`

```
$ elyra-metadata export component-catalogs --clean --directory=/xyz/tmp/foo3 --valid-only                  
[W 2022-02-15 08:11:05,425.425] Fetch of instance 'airflow_examples' of schemaspace 'component-catalogs' encountered an exception: No such schema named 'elyra-airflow-examples-catalog' was found in the component-catalogs schemaspace.
Creating directory structure for '/xyz/tmp/foo3/component-catalogs'
Traceback (most recent call last):
  File "/opt/anaconda3/envs/cli-metadata-export/bin/elyra-metadata", line 8, in <module>
    sys.exit(MetadataApp.main())
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 603, in main
    elyra_metadata.start()
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 624, in start
    return subinstance.start()
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 524, in start
    return subinstance.start()
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/site-packages/elyra/metadata/metadata_app.py", line 442, in start
    os.makedirs(dest_directory)
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/opt/anaconda3/envs/cli-metadata-export/lib/python3.7/os.py", line 223, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/xyz'

```

kevin-bates commented:
> @kevin-bates do I misremember or isn't there already code in place that should prevent commands other than migrate from running for schemaspaces that are deprecated?

We currently prevent 'install' operations on deprecated schemaspaces.
https://github.com/elyra-ai/elyra/blob/df9851243edceab90c9247b5a511fa706483c76b/elyra/metadata/metadata_app.py#L485

I would say that we _should_ support the ability to `export` deprecated schemaspace instances, but not `import` them.  Their export would then serve as a way to back them up prior to their deletion, for example.  In the previous example, do the 'aa' and 'kfp' instances get exported if `--valid-only` is removed?

ptitzler commented:
> I would say that we should support the ability to export deprecated schemaspace instances, but not import them.

+1

kevin-bates reviewed (approved):
This looks good @sandhya-nayak - thank you!

kevin-bates commented:
hmm - just noticed the test failure that I don't see when running locally.  Since this is one failure in export, we should hold off the merge until we've had a closer look.

sandhya-nayak commented:
> hmm - just noticed the test failure that I don't see when running locally. Since this is one failure in export, we should hold off the merge until we've had a closer look.
> 
> Looks like MacOS uses `mkdir: /xyz: Read-only file system`, but Linux uses: `mkdir: cannot create directory ‘/xyz’: Permission denied` so I think the assertion needs to be loosened up a bit.

@kevin-bates yeah - I've been trying to get that fixed - but I believe it isn't just a matter of loosening the assertion. I don't think the mocking is working at all (as per the "Interaction with mocking" section in the [pytest-console-scripts documentation](https://pypi.org/project/pytest-console-scripts/))

I tried making changes as per the suggestion in the documentation, but I can't get it to work.

So yeah - one way out is to get rid of the mocking altogether and attempt exporting to the /dummy_directory location and loosening the assertion. Only that isn't really a great solution because the test may still not work the same on, say, a Windows machine.

Thoughts? @ptitzler @akchinSTC

kevin-bates commented:
Yeah, I'm not sure how to trigger the `inprocess` mode that the referenced link talks about for mocking to work.  Its definitely running in `subprocess` mode: `elyra/tests/metadata/test_metadata_app.py::test_export_inaccessible_directory[subprocess]`

I think we should probably drop the use of mocking here and relax the assertion.  You're right, Windows will be a third flavor (although Elyra doesn't really run well on Windows to begin with).  I suppose you could assert that one of `Permission denied` or `Read-only file system` is also in the error message.  At least that way you know the error that occurs is related to what is expected.  I think Windows uses the term `Access Denied` so you could probably go ahead and include that in the list as well.

sandhya-nayak commented on elyra/metadata/metadata_app.py:
```diff
@@ -376,6 +378,100 @@ def start(self):
+            except OSError as e:
+                print(f"Error creating directory structure for {dest_directory}: {e.strerror}: '{e.filename}'")
+                self.exit(1)
+        else:
```
Moved the `if clean` portion to the `else` as there is nothing there to clean in a newly created directory

kevin-bates reviewed (approved):
Looks good @sandhya-nayak - thank you!

sandhya-nayak commented:
@ptitzler please review and let me know if anything else is needed here.

The updates include:
1. OSError handling + test for the same
2. Fixed typos/comments as per review comments
3. Summarized output (number of files exported instead of listing them out) + updated tests

ptitzler reviewed (approved):
LGTM! Very well done

ptitzler commented on elyra/metadata/metadata_app.py:
```diff
@@ -376,6 +378,100 @@ def start(self):
+                print(f"Creating directory structure for '{dest_directory}'")
+                os.makedirs(dest_directory)
+            except OSError as e:
+                print(f"Error creating directory structure for {dest_directory}: {e.strerror}: '{e.filename}'")
```
Nit

```suggestion
                print(f"Error creating directory structure for '{dest_directory}': {e.strerror}: '{e.filename}'")
```

ptitzler commented on elyra/metadata/manager.py:
```diff
@@ -87,7 +87,9 @@ def get_all(self, include_invalid: bool = False, of_schema: str = None) -> List[
-                if include_invalid:
+                if include_invalid and (not of_schema or invalid_instance.schema_name == of_schema):
+                    # Export invalid instances if requested and if a schema was not specified
+                    # or the speficied schema matches the instance's schema.
```
Typo:

```suggestion
                    # or the specified schema matches the instance's schema.
```

sandhya-nayak replied:
Thanks for catching this! Fixed it now.
